core/templates/dev/head/pages/exploration_player/PlaythroughService.js

```diff
@@ -0,0 +1,226 @@
+        };
+        return;
+      }
+
+      // Check for cyclic state transitions issue.
+      if (cycleIdentifier.num_cycles >= NUM_REPEATED_CYCLES_THRESHOLD) {
+        playthrough.issueType = ISSUE_TYPE_CYCLIC_STATE_TRANSITIONS;
+        playthrough.issueCustomizationArgs = {
+          state_names: {
+            value: cycleIdentifier.cycle.split(',')
+          }
+        };
+        return;
+      }
+
+      // Check for early quit issue.
+      if (timeSpentInExpInSecs < EARLY_QUIT_THRESHOLD_IN_SECS) {
+        playthrough.issueType = ISSUE_TYPE_EARLY_QUIT;
+        playthrough.issueCustomizationArgs = {
+          state_name: {
+            value:
+              playthrough.actions[
+                playthrough.actions.length - 1].actionCustomizationArgs
+                .state_name.value
+          },
+          time_spent_in_exp_in_secs: {
+            value: expStopwatch.getTimeInSecs()
+          }
+        };
+        return;
+      }
+    };
+
+    var getFullPlaythroughUrl = function() {
+      return UrlInterpolationService.interpolateUrl(
+        STORE_PLAYTHROUGH_URL, {
+          exploration_id: playthrough.expId
+        });
+    };
+
+    return {
+      initSession: function(newExplorationId, newExplorationVersion) {
+        playthrough = new PlaythroughObjectFactory(
+          null, newExplorationId, newExplorationVersion, null, {}, []);
+        expStopwatch = StopwatchObjectFactory.create();
+      },
+      getPlaythrough: function() {
+        return playthrough;
+      },
+      recordExplorationStartAction: function(initStateName) {
+        playthrough.actions.push(new LearnerActionObjectFactory(
+          ACTION_TYPE_EXPLORATION_START,
+          {
+            state_name: {
+              value: initStateName
+            }
+          },
+          CURRENT_ACTION_SCHEMA_VERSION
+        ));
+
+        multipleIncorrectStateName = {
+          state_name: initStateName,
+          num_times_incorrect: 0
+        };
+
+        cycleIdentifier = {
+          cycle: '',
+          num_cycles: 0
+        };
+        visitedStates.push(initStateName);
+      },
+      recordAnswerSubmitAction: function(
+          stateName, destStateName, interactionId, answer, feedback,
+          timeSpentInStateSecs) {
+        playthrough.actions.push(new LearnerActionObjectFactory(
+          ACTION_TYPE_ANSWER_SUBMIT,
+          {
+            state_name: {
+              value: stateName
+            },
+            dest_state_name: {
+              value: destStateName
+            },
+            interaction_id: {
+              value: interactionId
+            },
+            answer: {
+              value: answer
+            },
+            feedback: {
+              value: feedback
+            },
+            time_spent_in_state_secs: {
+              value: timeSpentInStateSecs
+            }
+          },
+          CURRENT_ACTION_SCHEMA_VERSION
+        ));
+
+        if (destStateName === stateName) {
```

Comment:
Please refactor these into functions, something like:

```javascript
var mISHandler = multipleIncorrectStatesHandlers['recordAnswerSubmitAction'];
if (mISHandler) mISHandler(...);
```

Reply:
@brianrodri I don't think I follow what you're suggesting. Could you please explain a bit more? Thanks.

Reply:
Done.